horovod/common/operations.cc

```diff
@@ -1993,27 +2113,92 @@ bool RunLoopOnce(HorovodGlobalState& state, bool is_coordinator) {
-            tensor_size += new_tensor_size;
-            response.add_tensor_names(new_response.tensor_names()[0]);
-            responses.pop_front();
-          } else {
-            // Don't try to fuse additional tensors since they are usually
-            // computed in order of requests and skipping tensors may mean
-            // that the batch will have to wait longer while skipped tensors
-            // could be reduced at that time.
-            break;
+            if (response.response_type() == new_response.response_type() &&
+                response.devices() == new_response.devices() &&
+                entry.tensor->dtype() == new_entry.tensor->dtype() &&
+                tensor_size + new_tensor_size <= TensorFusionThresholdBytes()) {
+              // These tensors will fuse together well.
+              tensor_size += new_tensor_size;
+              response.add_tensor_names(new_response.tensor_names()[0]);
+              responses.pop_front();
+            } else {
+              // Don't try to fuse additional tensors since they are usually
+              // computed in order of requests and skipping tensors may mean
+              // that the batch will have to wait longer while skipped tensors
+              // could be reduced at that time.
+              break;
+            }
           }
-        }
-      }
+        } else if (response.response_type() ==
+                   MPIResponse::ResponseType::ALLGATHER) {
+          // Attempt to add more responses to this fused response.
+          auto& entry = state.tensor_table[response.tensor_names()[0]];
 
+          // Copy tensor sizes from the MPI response into a vector of int64_t
+          // and compute total size.  This is size of first dimension.
+          int64_t total_dimension_size =
+              TotalDimensionSize(response.tensor_sizes());
+
+          // Every tensor participating in Allgather operation may have
+          // different first dimension size, but the rest of dimensions are same
+          // for all tensors.  Here we get shape of tensor sliced by first
+          // dimension. Allgather output will have shape of: (sum of first
+          // dimension of every tensor) x (tensor slice shape).
+          int64_t total_count_of_output_entries = total_dimension_size;
+          for (int i = 1; i < entry.tensor->shape().dims(); ++i) {
+            total_count_of_output_entries *= entry.tensor->shape().dim_size(i);
+          }
+          int element_size;
+          MPI_Type_size(GetMPIDataType(entry.tensor), &element_size);
+          int64_t total_byte_size_of_output =
+              total_count_of_output_entries * element_size;
+
+          while (!responses.empty()) {
+
+            auto new_response = responses.front();
+            assert(new_response.tensor_names().size() == 1);
+            auto& new_entry =
+                state.tensor_table[new_response.tensor_names()[0]];
+
+            int64_t new_total_dimension_size =
```

Comment:
Can you refactor it to return `total_count_of_output_entries`?  That's the piece of computation you're duplicating as well.